Index: mt76-2021-12-03-678071ef/mt7915/init.c
===================================================================
--- mt76-2021-12-03-678071ef.orig/mt7915/init.c
+++ mt76-2021-12-03-678071ef/mt7915/init.c
@@ -512,6 +512,8 @@ static void mt7915_wfsys_reset(struct mt
 #define MT_MCU_DUMMY_RANDOM	GENMASK(15, 0)
 #define MT_MCU_DUMMY_DEFAULT	GENMASK(31, 16)
 
+	dev_err(dev->mt76.dev, "resetting chip\n");
+
 	mt76_wr(dev, MT_MCU_WFDMA0_DUMMY_CR, MT_MCU_DUMMY_RANDOM);
 
 	/* change to software control */
@@ -553,10 +555,7 @@ static int mt7915_init_hardware(struct m
 	INIT_WORK(&dev->init_work, mt7915_init_work);
 	dev->dbdc_support = !!(mt76_rr(dev, MT_HW_BOUND) & BIT(5));
 
-	/* If MCU was already running, it is likely in a bad state */
-	if (mt76_get_field(dev, MT_TOP_MISC, MT_TOP_MISC_FW_STATE) >
-	    FW_STATE_FW_DOWNLOAD)
-		mt7915_wfsys_reset(dev);
+	mt7915_wfsys_reset(dev);
 
 	ret = mt7915_dma_init(dev);
 	if (ret)
