From 4d7e38d8da39c59d3e91ad114d307776b4d5d3d5 Mon Sep 17 00:00:00 2001
From: Chad Monroe <chad.monroe@smartrg.com>
Date: Thu, 14 Apr 2022 11:49:29 -0700
Subject: [PATCH 2/2] mt76: mt7915: disable TWT support

---
 mt7915/init.c | 2 --
 mt7915/main.c | 2 --
 2 files changed, 4 deletions(-)

--- a/mt7915/init.c
+++ b/mt7915/init.c
@@ -924,8 +924,6 @@ mt7915_init_he_caps(struct mt7915_phy *p
 
 		switch (i) {
 		case NL80211_IFTYPE_AP:
-			he_cap_elem->mac_cap_info[0] |=
-				IEEE80211_HE_MAC_CAP0_TWT_RES;
 			he_cap_elem->mac_cap_info[2] |=
 				IEEE80211_HE_MAC_CAP2_BSR;
 			he_cap_elem->mac_cap_info[4] |=
--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -1489,8 +1489,6 @@ const struct ieee80211_ops mt7915_ops =
 	.sta_statistics = mt7915_sta_statistics,
 	.sta_set_4addr = mt7915_sta_set_4addr,
 	.sta_set_decap_offload = mt7915_sta_set_decap_offload,
-	.add_twt_setup = mt7915_mac_add_twt_setup,
-	.twt_teardown_request = mt7915_twt_teardown_request,
 	CFG80211_TESTMODE_CMD(mt76_testmode_cmd)
 	CFG80211_TESTMODE_DUMP(mt76_testmode_dump)
 #ifdef CONFIG_MAC80211_DEBUGFS
