Index: mt76-2022-04-28-784c27b1/mt7915/init.c
===================================================================
--- mt76-2022-04-28-784c27b1.orig/mt7915/init.c
+++ mt76-2022-04-28-784c27b1/mt7915/init.c
@@ -582,7 +582,7 @@ static void mt7915_init_work(struct work
 	mt7915_init_txpower(dev, &dev->mphy.sband_5g.sband);
 	mt7915_init_txpower(dev, &dev->mphy.sband_6g.sband);
 	mt7915_txbf_init(dev);
-	dev->dbg.muru_onoff = OFDMA_DL | MUMIMO_UL | MUMIMO_DL;
+	dev->dbg.muru_onoff = 0;
 }
 
 void mt7915_wfsys_reset(struct mt7915_dev *dev)
